deepmd/pt/model/descriptor/repflow_layer.py

```diff
@@ -0,0 +1,938 @@
+    ):
+        """
+        Parameters
+        ----------
+        node_ebd_ext : nf x nall x n_dim
+            Extended node embedding.
+        edge_ebd : nf x nloc x nnei x e_dim
+            Edge embedding.
+        h2 : nf x nloc x nnei x 3
+            Pair-atom channel, equivariant.
+        angle_ebd : nf x nloc x a_nnei x a_nnei x a_dim
+            Angle embedding.
+        nlist : nf x nloc x nnei
+            Neighbor list. (padded neis are set to 0)
+        nlist_mask : nf x nloc x nnei
+            Masks of the neighbor list. real nei 1 otherwise 0
+        sw : nf x nloc x nnei
+            Switch function.
+        a_nlist : nf x nloc x a_nnei
+            Neighbor list for angle. (padded neis are set to 0)
+        a_nlist_mask : nf x nloc x a_nnei
+            Masks of the neighbor list for angle. real nei 1 otherwise 0
+        a_sw : nf x nloc x a_nnei
+            Switch function for angle.
+
+        Returns
+        -------
+        n_updated:     nf x nloc x n_dim
+            Updated node embedding.
+        e_updated:     nf x nloc x nnei x e_dim
+            Updated edge embedding.
+        a_updated : nf x nloc x a_nnei x a_nnei x a_dim
+            Updated angle embedding.
+        """
+        nb, nloc, nnei, _ = edge_ebd.shape
+        nall = node_ebd_ext.shape[1]
+        node_ebd = node_ebd_ext[:, :nloc, :]
+        assert (nb, nloc) == node_ebd.shape[:2]
+        assert (nb, nloc, nnei) == h2.shape[:3]
+        del a_nlist  # may be used in the future
+
+        n_update_list: list[torch.Tensor] = [node_ebd]
+        e_update_list: list[torch.Tensor] = [edge_ebd]
+        a_update_list: list[torch.Tensor] = [angle_ebd]
+
+        # node self mlp
+        node_self_mlp = self.act(self.node_self_mlp(node_ebd))
+        n_update_list.append(node_self_mlp)
+
+        nei_node_ebd = _make_nei_g1(node_ebd_ext, nlist)
+
+        # node sym (grrg + drrd)
+        node_sym_list: list[torch.Tensor] = []
+        node_sym_list.append(
+            self.symmetrization_op(
+                edge_ebd,
+                h2,
+                nlist_mask,
+                sw,
+                self.axis_neuron,
+            )
+        )
+        node_sym_list.append(
+            self.symmetrization_op(
+                nei_node_ebd,
+                h2,
+                nlist_mask,
+                sw,
+                self.axis_neuron,
+            )
+        )
+        node_sym = self.act(self.node_sym_linear(torch.cat(node_sym_list, dim=-1)))
+        n_update_list.append(node_sym)
+
+        if not self.optim_update:
+            # nb x nloc x nnei x (n_dim * 2 + e_dim)
+            edge_info = torch.cat(
+                [
+                    torch.tile(node_ebd.unsqueeze(-2), [1, 1, self.nnei, 1]),
+                    nei_node_ebd,
+                    edge_ebd,
+                ],
+                dim=-1,
+            )
+        else:
+            edge_info = None
```

Comment:
I kept the original update here for clarity and future optimization test. The default method is within `optim_update`. We can remove this branch if necessary in the future.